src/workerd/jsg/string.h

```diff
@@ -169,27 +169,27 @@ class UsvStringIterator {
   inline explicit operator bool() const { return pos < ptr.size(); }
 
   inline bool operator<(UsvStringIterator& other) const {
-    return pos < other.pos && ptr == other.ptr;
+    return pos < other.pos && ptr.begin() == other.ptr.begin();
```

Comment:
Since these operators are used for things like for-loop iteration, I guess it's typically the case that `ptr` and `other.ptr` are going to have exactly the same value, and it's going to be pretty rare that someone is comparing iterators from different underlying strings.  But if they did, I assume it's OK that we're switching to compare `ptr.begin()`, but not also `ptr.size()` -- i.e. comparisons where the underlying strings share the same starting position in memory but one is a substring of the other.

(Or maybe the check for the same underlying string should effectively be an assert in these comparisons? ...although I don't know if that itself will also negatively affect performance.)

Reply:
our use of these is pretty rare and limited to URL and URLPattern right now, and for those we know how they are being used, so this is safe for now and we can revisit if the use cases expand.

Reply:
The existing code actually compares the _content_ of the arrays, which I guess is a bug?

In general, comparing iterators from different containers is expected to be UB in C++. In fact, I believe it's technically UB to use `<` or `>` to compare pointers that do not point into the same underlying array. So it would be idiomatic C++ to remove these "same string" checks altogether.

Moreover, even if we wanted to allow comparison of iterators from different strings, I think these comparisons are incorrect as written, because they produce inconsistent orderings. That is, `i < j`, `j < i`, and `i == j` will all evaluate false when the iterators are from different strings. Most algorithms that use ordering probably won't work when presented with such inconsistency, and some standard library algorithms will even end up UB.

So I'd argue that we really should just delete these extra checks and go with the standard expectation that you cannot compare iterators from different containers at all.